reversible migrations for cli

sidred opened the pull request:
- add -r flag to migration add command which if true will generate an up and a down file
- add revert subcommand, which reverts the last migration
- add --dry-run flag to migration run command, which list the migrations that will be applied
- updates add migration to check if all migration are of same type, i.e cannot mix and match reversible and simple migrations
- add database reset subcommand which drops and creates database and applies migrations

Tested on Postgresql. Need to test on mysql and sqlite

jplatte commented on sqlx-cli/src/migrate.rs:
```diff
@@ -28,7 +30,30 @@ pub fn add(description: &str) -> anyhow::Result<()> {
+    fs::create_dir_all(MIGRATION_FOLDER).context("Unable to create migrations directory")?;
+
+    let migrator = Migrator::new(Path::new(MIGRATION_FOLDER)).await?;
+    // This checks if all existing migrations are of the same type as the reverisble flag passed
```
```suggestion
    // This checks if all existing migrations are of the same type as the reversible flag passed
```

sidred commented:
Some of the tests are failing because of error in the migrate! macro. The error seems to originate in Migrate constructor in QuotedMigration ToToken implementation.

https://github.com/launchbadge/sqlx/blob/67099d993c5c9eb575ca530c0c94c9a9066c1ebd/sqlx-macros/src/migrate.rs#L27

However i am unable to pass migration_type in the constructor. I tried making the function `expand_migrator_from_dir` in the same file to match the changes i made in source.rs file and passing migration_type to QuotedMigration. However i get this error

`the trait bound `sqlx_core::migrate::migration_type::MigrationType: quote::to_tokens::ToTokens` is not satisfied`

Any help on how to resolve this is appreciated.

raftario commented:
The easiest way to fix it would be to create a new `QuotedMigrationType` newtype and implement `ToTokens` for it.
```rs
put struct QuotedMigrationType(MigrationType);
impl ToTokens for QuotedMigrationType {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let ts = match self.0 {
            MigrationType::Simple => quote! { sqlx::migrate::MigrationType::Simple },
            // ...
        };
        tokens.append_all(ts.into_iter());
    }
}
```

izik1 commented:
@sidred could I have you rebase?
My triage efforts would also appreciate it if you could tell me if you're waiting on a reviewer or being waited on

sidred commented:
I got busy with work, will try to rebase it this weekend, if someone else doesn't get to it

Pending tasks
- Figure out any tests to be added
- Tested only with postgres. Need to test other db's
- Remove reset command. since this functionality has been merge via another PR. 
- They were a few pending PR's for cli, figure out what other features if any have been merged.

mehcode commented:
@sidred Are you still interested in rebasing this? Very nice work.

sidred commented:
I have rebased with the latest master and removed the reset command from this PR which was already added by #668

Only pending item is the tests. Not sure how to go about writing test cases for this

mehcode commented:
Thank you @sidred , we can iterate on improving our test coverage
